### PR TITLE
Make meta a device (getting rid of empty_meta)

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53143 Make meta a device (getting rid of empty_meta)**
* #53142 Add NoOpDeviceGuardImpl
* #53116 Delete dead Backend toSparse

Meta is now an honest to goodness device type, like cpu, so you can use
device='meta' to trigger allocation of meta tensors.  This way better
than empty_meta since we now have working API for most factory functions
(they don't necessarily work yet, though, because need to register Meta
versions of those functions.)

Some subtleties:
- I decided to drop the concept of CPU versus CUDA meta tensors; meta
  tensors are device agnostic.  It's hard to say exactly what the
  correct level of abstraction here is, but in this particular case
  implementation considerations trump semantic considerations: it
  is way easier to have just a meta device, than to have a meta device
  AND a cpu device AND a cuda device.  This may limit the applicability
  of meta tensors for tracing models that do explicit cpu()/cuda()
  conversions (unless, perhaps, we make those operations no-ops on meta
  tensors).
- I noticed that the DeviceType uppercase strings are kind of weird.
  Are they really supposed to be all caps?  That's weird.
- I moved the Meta dispatch key to live with the rest of the "device"
  dispatch keys.
- I intentionally did NOT add a Backend for Meta.  For now, I'm going to
  hope meta tensors never exercise any of the Backend conversion code;
  even if it does, better to fix the code to just stop converting to and
  from Backend.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26763552](https://our.internmc.facebook.com/intern/diff/D26763552)

### CODE REVIEW

**Comment by ezyang** [comment]
Distributed, y'all got tagged because I updated your test on the error message to be less brittle to device list changes

File: c10/core/DispatchKey.h [hunk 1]
```diff
@@ -76,6 +76,13 @@ enum class DispatchKey : uint8_t {
   OpenCL,
   IDEEP,
 
+  // A meta tensor is a tensor without any data associated with it.  (They
+  // have also colloquially been referred to as tensors on the "null" device).
+  // A meta tensor can be used to dry run operators without actually doing any
+  // computation, e.g., add on two meta tensors would give you another meta
+  // tensor with the output shape and dtype, but wouldn't actually add anything.
+  Meta,
```
Comment (bdhirsh) [inline c10/core/DispatchKey.h:84]: It seems like the old comment here tried to explain the differences between a meta tensor and a meta function (which is called by both normal and meta tensors), which I thought was pretty helpful context.

File: c10/core/DispatchKey.h [hunk 2]
```diff
@@ -123,58 +130,6 @@ enum class DispatchKey : uint8_t {
   // If you add new backend keys after PrivateUse3, please also update it here.
   EndOfBackendKeys = PrivateUse3,
```
Comment (bdhirsh) [inline c10/core/DispatchKey.h:131]: Just thinking out loud- is the Meta key above this so that's is a "real" backend just a semantic change? Or does it actually change behavior. I think the answer is that it doesn't change behavior- since Meta was already lower priority than the other non-backend keys, and it doesn't redispatch.

File: c10/core/TensorOptions.h [hunk 3]
```diff
@@ -691,6 +693,8 @@ inline DeviceType computeDeviceType(DispatchKey tid) {
     return DeviceType::XLA;
   } else if (tid == DispatchKey::MLC) {
     return DeviceType::MLC;
+  } else if (tid == DispatchKey::Meta) {
+    return DeviceType::Meta;
```
Comment (bdhirsh) [inline c10/core/TensorOptions.h:697]: I guess this change *technically* makes meta computations marginally slower, since now they have to go through BackendSelect to get to the meta kernel. Probably fine? Especially if we're hoping to improve BackendSelect perf soon anyway.

File: torch/testing/_internal/distributed/nn/api/remote_module_test.py [hunk 4]
```diff
@@ -257,8 +257,7 @@ def test_invalid_devices(self):
 
         with self.assertRaisesRegex(
             RuntimeError,
-            r"Expected one of cpu, cuda, xpu, mkldnn, opengl, opencl, ideep, hip, msnpu, mlc, xla, vulkan"
-            " device type at start of device string",
+            r"Expected one of .+ device type at start of device string",
```
Comment (bdhirsh) [inline torch/testing/_internal/distributed/nn/api/remote_module_test.py:260]: Is this related?

Unless this is somehow regex-filling-in all possible device types and I don't see how 😛

**Review by bdhirsh — APPROVED** [review]
> it is way easier to have just a meta device, than to have a meta device
AND a cpu device AND a cuda device. This may limit the applicability
of meta tensors for tracing models that do explicit cpu()/cuda()
conversions (unless, perhaps, we make those operations no-ops on meta
tensors).

It sounds like we maybe shouldn't bother with this unless we find out later that it'd be really useful. But I could imagine adding an `is_meta` boolean to `Device` could give us all of that without complicating the design too much? At the downside of probably making `sizeof(Device)` a little bigger, since I think it's 16 bytes right now.

File: c10/core/DispatchKey.h (hunk 1 above)
Comment (ezyang) [reply c10/core/DispatchKey.h:84]: The comment was hopelessly out of date though :) I can bring back some new text explaining the difference between a meta function and a meta tensor, though

File: c10/core/DispatchKey.h (hunk 2 above)
Comment (ezyang) [reply c10/core/DispatchKey.h:131]: It shouldn't be a behavior change, because (I think) there aren't any interactions between Meta and the keys I moved it past. Lack of redispatch from Meta helps a bit.

File: c10/core/TensorOptions.h (hunk 3 above)
Comment (ezyang) [reply c10/core/TensorOptions.h:697]: Also I don't care about Meta perf ;)

File: torch/testing/_internal/distributed/nn/api/remote_module_test.py (hunk 4 above)
Comment (ezyang) [reply torch/testing/_internal/distributed/nn/api/remote_module_test.py:260]: Yeah, I changed the list of valid device types, so this test is wrong. I just got rid of the list match entirely and replaced it with a wildcard.

**Comment by ezyang** [comment]
> But I could imagine adding an is_meta boolean to Device could give us all of that without complicating the design too much? 

It's an interesting thought. One of the reasons I'm concerned is I am not sure Meta is going to be the last of its kind. If we do lazy initialization with lazy tensors, that's gonna be another thing: can you have lazy cpu and lazy cuda? Big can of worms.

**Comment by bdhirsh** [comment]
Oh also- does `.to(device)` play nicely with meta out of the box? I think you might run into issues if you call:
```
a = torch.empty(1, device='meta')
b = a.to('cpu')
c = b + torch.ones(1) # we probably want b to have a storage at this point, but I don't think it will
c = a + torch.ones(1) # the first argument has device 'meta`' and the the second is on cpu- probably error here?
```

**Comment by ezyang** [comment]
> Oh also- does .to(device) play nicely with meta out of the box? I think you might run into issues if you call:

Right now it fails because there is no Meta kernel associated with to. A good question is what it should do. There are a few possibilities:

1. It continues to error
2. It allocates a CPU tensor with the appropriate size per the meta tensor, but uninitialized contents
3. It gives you back another meta tensor

**Comment by mgoldenbe** [comment]
I seem to encounter a related error. Please see my [question at StackOverflow](https://stackoverflow.com/q/76883181/2725810).
